File: src/main/scala/com/github/chengpohi/script/ScriptContext.scala
package com.github.chengpohi.script

import com.github.chengpohi.context._
import com.github.chengpohi.edql.parser.json.JsonCollection
import org.apache.commons.codec.binary.Base64

import scala.collection.mutable
import scala.concurrent.duration
import scala.concurrent.duration.Duration


sealed case class ScriptContext(hostInfo: HostInfo) extends EDQLConfig with Context {
  override implicit val resultTimeout: Duration = Duration.apply(hostInfo.timeout, duration.MILLISECONDS)
  override implicit val kibanaProxy: Boolean = hostInfo.kibanaProxy
  override implicit val readOnly: Boolean = hostInfo.readOnly
  override implicit lazy val eqlClient: EDQLClient = buildRestClient(hostInfo)

  def clear(): Unit = variables.clear()
}

object ScriptContext {
  val cache: mutable.Map[String, (Long, ScriptContext)] = mutable.Map[String, (Long, ScriptContext)]()

  def apply(hostInfo: HostInfo,
            vars: Map[String, JsonCollection.Val]): ScriptContext = {
    val cacheInfo = getCacheConnection(hostInfo)
    val cacheKey = cacheInfo._1
    val cacheContext = cacheInfo._2
    if (cacheContext.isDefined) {
      val c = cacheContext.get._2
      c.variables = mutable.Map[String, JsonCollection.Val](vars.toSeq: _*)
      return c
    }

    val context = new ScriptContext(hostInfo)
    context.variables = mutable.Map[String, JsonCollection.Val](vars.toSeq: _*)
    cache.put(cacheKey, (System.currentTimeMillis(), context))
    context
  }

  def getCacheConnection(hostInfo: HostInfo) = {
    val cacheKey =
      s"""${hostInfo.host}-${hostInfo.authInfo.map(i => i.cacheKey).getOrElse("")}
         |-${hostInfo.timeout}-${hostInfo.kibanaProxy}-${hostInfo.readOnly}-
         |${hostInfo.proxyInfo.map(i => i.cacheKey).getOrElse("")}""".stripMargin

    val cacheContext = cache.get(cacheKey)
    cacheContext match {
      case None => (cacheKey, None)
      case Some(c) => {
        c._2.eqlClient.restClient.isRunning match {
          case true => (cacheKey, cacheContext)
          case false => (cacheKey, None)
        }
      }
    }
  }

  def removeCacheConnection(hostInfo: HostInfo) = {
    val keyBytes = (s"$hostInfo.endpoint-" + hostInfo.authInfo.map(i => i.cacheKey).getOrElse("")
      + s"-${hostInfo.timeout}" + s"-${hostInfo.kibanaProxy}" + s"-${hostInfo.readOnly}"
      + s"-${hostInfo.proxyInfo.map(i => i.cacheKey).getOrElse("")}").getBytes

    val cacheKey = Base64.encodeBase64String(keyBytes)

    cache.remove(cacheKey)
  }
}

File: src/main/java/com/github/chengpohi/script/ScriptContext.kt
package com.github.chengpohi.script

import com.github.chengpohi.context.HostInfo
import com.github.chengpohi.edql.parser.json.JsonCollection
import org.apache.commons.codec.binary.Base64
import java.util.concurrent.ConcurrentHashMap
import kotlin.time.Duration
import kotlin.time.Duration.Companion.milliseconds

class ScriptContext(val hostInfo: HostInfo) {
    val client: EDQLClient by lazy { EDQLClient(hostInfo) }
    var variables: MutableMap<String, JsonCollection.Val> = mutableMapOf()
    val timeout: Duration = hostInfo.timeout.milliseconds
    val kibanaProxy: Boolean = hostInfo.kibanaProxy
    val readOnly: Boolean = hostInfo.readOnly
    val pathPrefix: String = hostInfo.uri.path

    fun clear() {
        variables.clear()
    }

    companion object {
        val cache: MutableMap<String, Pair<Long, ScriptContext>> = ConcurrentHashMap()

        fun apply(
            hostInfo: HostInfo,
            vars: Map<String, JsonCollection.Val>
        ): ScriptContext {
            val cacheInfo = getCacheConnection(hostInfo)
            val cacheKey = cacheInfo.first
            val cacheContext = cacheInfo.second
            if (cacheContext != null) {
                val c = cacheContext.second
                c.variables = vars.toMutableMap()
                return c
            }

            val context = ScriptContext(hostInfo)
            context.variables = vars.toMutableMap()
            cache[cacheKey] = Pair(System.currentTimeMillis(), context)
            return context
        }

        fun getCacheConnection(hostInfo: HostInfo): Pair<String, Pair<Long, ScriptContext>?> {
            val cacheKey = buildString {
                append(hostInfo.host)
                append("-")
                append(hostInfo.authInfo?.cacheKey() ?: "")
                append("-")
                append(hostInfo.timeout)
                append("-")
                append(hostInfo.kibanaProxy)
                append("-")
                append(hostInfo.readOnly)
                append("-")
                append(hostInfo.proxyInfo?.cacheKey() ?: "")
            }

            val cacheContext = cache[cacheKey]
            return when {
                cacheContext == null -> Pair(cacheKey, null)
                cacheContext.second.client.isRunning() -> Pair(cacheKey, cacheContext)
                else -> Pair(cacheKey, null)
            }
        }

        fun removeCacheConnection(hostInfo: HostInfo) {
            val keyBytes = buildString {
                append(hostInfo)
                append(".endpoint-")
                append(hostInfo.authInfo?.cacheKey() ?: "")
                append("-")
                append(hostInfo.timeout)
                append("-")
                append(hostInfo.kibanaProxy)
                append("-")
                append(hostInfo.readOnly)
                append("-")
                append(hostInfo.proxyInfo?.cacheKey() ?: "")
            }.toByteArray()

            val cacheKey = Base64.encodeBase64String(keyBytes)
            cache.remove(cacheKey)
        }
    }
} 